staking/staking_hub.go

```diff
@@ -0,0 +1,409 @@
+)
+
+var (
+	defaultFileName = "staking-map.json"
+)
+
+// StakingHub acts as a hub (manager) for staked account balances
+type StakingHub struct {
+	// Address -> Stake
+	StakingMap map[types.Address]*big.Int
+
+	// Specifies the working directory for the Polygon SDK (location for writeback)
+	WorkingDirectory string
+
+	// Write-back period (in s) for backup staking data
+	WritebackPeriod time.Duration
+
+	// Logger for logging errors and information
+	Logger hclog.Logger
+
+	// Event queue defines staking / unstaking events
+	// which are read by modules that do final transaction sealing
+	EventQueue []PendingEvent
+
+	// Event list mutex
+	EventQueueMutex sync.Mutex
+
+	// Staking map mutex
+	StakingMutex sync.Mutex
+
+	// Close channel
+	CloseCh chan struct {
+	}
+}
+
+var stakingHubInstance StakingHub
+var once sync.Once
+
+// GetStakingHub initializes the stakingHubInstance singleton
+func GetStakingHub() *StakingHub {
+	once.Do(func() {
+		stakingHubInstance = StakingHub{
+			StakingMap:      make(map[types.Address]*big.Int),
+			EventQueue:      make([]PendingEvent, 0),
+			CloseCh:         make(chan struct{}),
+			WritebackPeriod: 60 * time.Second,
+		}
+	})
+
+	return &stakingHubInstance
+}
+
+// SetWorkingDirectory sets the writeback directory for the staking map
+func (sh *StakingHub) SetWorkingDirectory(directory string) {
+	sh.WorkingDirectory = directory
+
+	err := sh.readFromDisk()
+	if err != nil {
+		// Log as warning because reading this map is not crucial
+		sh.log(err.Error(), logWarning)
+	}
+
+	go sh.saveToDisk()
+}
+
+// LogType defines possible log types for the logger
+type LogType string
+
+var (
+	logInfo    LogType = "info"
+	logError   LogType = "error"
+	logWarning LogType = "warning"
+)
+
+// log logs the output to the console if the logger is set
+func (sh *StakingHub) log(output string, logType LogType) {
+	if sh.Logger != nil {
+		switch logType {
+		case logInfo:
+			sh.Logger.Info(output)
+		case logError:
+			sh.Logger.Error(output)
+		case logWarning:
+			sh.Logger.Warn(output)
+		}
+	}
+}
+
+// SetLogger sets the StakingHub logger
+func (sh *StakingHub) SetLogger(logger hclog.Logger) {
+	sh.Logger = logger.Named("staking-hub")
+}
+
+// CloseStakingHub stops the writeback process
+func (sh *StakingHub) CloseStakingHub() {
+	sh.StakingMutex.Lock()
+	defer sh.StakingMutex.Unlock()
+
+	// Alert the closing channel
+	sh.CloseCh <- struct{}{}
+
+	close(sh.CloseCh)
+}
+
+// PendingEvent contains useful information about a staking / unstaking event
+type PendingEvent struct {
+	BlockNumber int64
+	Address     types.Address
+	Value       *big.Int
+	EventType   StakingEventType
+}
+
+// Compare checks if the two events match
+func (pe *PendingEvent) Compare(event PendingEvent) bool {
+	if pe.EventType == event.EventType &&
+		pe.Address.String() == event.Address.String() &&
+		pe.Value.Cmp(event.Value) == 0 &&
+		pe.BlockNumber == event.BlockNumber {
+		return true
+	}
+
+	return false
+}
```

Comment:
The implementation assumed that the user cannot / should not send 2+ staking / unstaking transactions within a single block. I believe our IBFT validator event logic is based on this as well.

The only thing that is missing from the solution for it to respect the point stated above is to deny the user staking if he has tried to stake in the same block in `system_staking.go`.

Reply:
@Vuksan For some reason I cannot reply to [this comment](https://github.com/0xPolygon/polygon-sdk/pull/84#discussion_r679095525) directly, so I'll write the reply here.

I'll modify the description a bit, so it's more clear as to what the method returns and at which point in time 👍

Reply:
> The implementation assumed that the user cannot / should not send 2+ staking / unstaking transactions within a single block. I believe our IBFT validator event logic is based on this as well.
> 
> The only thing that is missing from the solution for it to respect the point stated above is to deny the user staking if he has tried to stake in the same block in `system_staking.go`.
> 
> @lazartravica @Kourin1996 correct me if I'm wrong on this assumption

Ignore this comment completely, we do have support for multiple staking / unstaking transactions within a single block. The IBFT validator event logic is executed after things have been committed 👍 

@Vuksan Adding the ID to differentiate the events further. Good catch 👍

Reply:
I decided it didn't make sense to add the ID, as these events are constructed on the fly both in the staking / unstaking handles as well as in the executor when going over transactions.

The fix was simply to remove the `hasEvent` check when adding the pending event, because that's what prevented double adds, and it was leftover code to patch the proposer double event add - and has now been resolved through other code flows 👍

Reply:
Alright, but now we would have to instruct the caller of the `RemovePendingEvent` to call it multiple times if he wishes to remove a specific event, because multiple instances of the same event may be present.

I think that having that ID would clear any doubts for the caller, which will prove beneficial in the long run.

Reply:
I'm not sure I understand why you would need to call `RemovePendingEvent` manually.
All events will eventually be removed / added if they are in the queue. 

The problem with adding an ID is that you have no idea what kind of ID was assigned to a specific event in the `system_staking` or `system_unstaking` handlers, while you're in the `executor` and constructing events for searches.

Could you please clarify this?

Reply:
> I'm not sure I understand why you would need to call RemovePendingEvent manually.
All events will eventually be removed / added if they are in the queue.

I don't think right now, but this is a public method. We may have numerous bugs tomorrow because we may rely on this method to tell us if the two events are the same event, while in reality it can lie to us stating that two events are the same one, although they are not.

> The problem with adding an ID is that you have no idea what kind of ID was assigned to a specific event in the system_staking or system_unstaking handlers, while you're in the executor and constructing events for searches.

I meant to have an integer ID that increases whenever a new event is created.
How come you don't know the ID? Wouldn't it be emitted as well, since it would be a part of the event structure?

Reply:
It's not a problem to add an ID in the form of a UUID, the issue arises in `executor.go` when you're parsing transactions and need to construct an event for the search:
https://github.com/0xPolygon/polygon-sdk/blob/37e054cef0a49d6669db45c59f1778473318e7c9/state/executor.go#L184-L189

and

https://github.com/0xPolygon/polygon-sdk/blob/37e054cef0a49d6669db45c59f1778473318e7c9/state/executor.go#L165-L170

At this point in time, you have no idea what the ID of the pending event that corresponds to that transaction that you're currently parsing is - which means if you don't know this ID, you won't be able to do a search on the event queue if the ID is part of that search.

The only solution I see to this is to add some kind of tracking mechanism in `executor.go` to know which event corresponds to which ID, but it would only complicate the code even more.

Reply:
@zivkovicmilos @Vuksan What do you think about using the transaction nonce instead of the ID?

Reply:
Oh, I get it now.
What do you think of using the `nonce` as @lazartravica suggested, of course it would have to be in combination with `From` I guess?